pkg/reconciler/instances/istio/reset/pod/handler.go

```diff
@@ -140,3 +183,132 @@ func doRollout(customObject CustomObject, kubeClient kubernetes.Interface) (err
 
 	return
 }
+
+func isDeploymentReady(deployment *v1.Deployment, kubeClient kubernetes.Interface) bool {
+	if deployment.DeletionTimestamp != nil {
+		return false
+	}
+	_, _, newReplicaSet, err := kctlutil.GetAllReplicaSets(deployment, kubeClient.AppsV1())
+	if err != nil || newReplicaSet == nil {
+		return false
+	}
+	if newReplicaSet.Status.ReadyReplicas < *deployment.Spec.Replicas {
+		return false
+	}
+
+	return true
+}
+
+func isStatefulSetReady(sts *v1.StatefulSet) bool {
+	if sts.DeletionTimestamp != nil {
+		return false
+	}
+
+	if sts.Generation <= sts.Status.ObservedGeneration {
+		if sts.Status.UpdateRevision != sts.Status.CurrentRevision {
+			return false
+		}
+		if sts.Status.ReadyReplicas != *sts.Spec.Replicas {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isDaemonSetReady(ds *v1.DaemonSet) bool {
+	if ds.DeletionTimestamp != nil {
+		return false
+	}
+
+	if ds.Generation <= ds.Status.ObservedGeneration {
+		if ds.Status.UpdatedNumberScheduled < ds.Status.DesiredNumberScheduled {
+			return false
+		}
+		if ds.Status.NumberAvailable < ds.Status.DesiredNumberScheduled {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isReplicaSetReady(rs *v1.ReplicaSet) bool {
+	if rs.DeletionTimestamp != nil {
+		return false
+	}
+	if rs.Status.Replicas != rs.Status.AvailableReplicas &&
+		rs.Status.Replicas != rs.Status.ReadyReplicas {
+		return false
+	}
+
+	return true
+}
+
+func (i *RolloutHandler) waitForDeployment(object CustomObject) error {
+	err := wait.Poll(i.waitOpts.Interval, i.waitOpts.Timeout, func() (done bool, err error) {
+		dep, err := i.kubeClient.AppsV1().Deployments(object.Namespace).Get(context.Background(), object.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		ready := isDeploymentReady(dep, i.kubeClient)
+		return ready, nil
+	})
+	if err != nil {
+		return err
+	}
+	i.log.Infof("%s/%s/%s is ready", object.Kind, object.Namespace, object.Name)
+
+	return nil
+}
```

Comment:
iiuc, `wait.Poll` seams to bring no new functionalities comparing to `retry` package that we are using in other cases. it also brought the `WaitOptions` without need imho. Would it be possible to use `retry` package here and not bringing another lib/pattern?

Reply:
Does bring the functionality to poll with the condition function.
Retry is used here for reliability of k8s API calls, but lacks the simplicity of polling the state with condition func with strict timeout configuration, which I feel is much better for the task such like this one.